---
title: "Nucleotide periodicity assessment of gene transcription and translation under stress in *H. sapiens* and *E. coli*"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
# Reads in D.E. Andreev data, and filters for the 10% FDR threshold for riboseq
# being True
library(readr)
df <- read_csv("./data/hsapiens/riboseq_mRNA_deandreev.csv")
df_tsl_resp <- filter(df, `exceeds 10% FDR threshold` == "True")

# Filters to just deal with ribo-seq based data
df_tsl_resp <- df_tsl_resp[,1:4]
glimpse(df_tsl_resp)

df_tsl_resp$`Gene symbol` <- toupper(df_tsl_resp$`Gene symbol`)

write_csv(select(df_tsl_resp, `Gene symbol`) ,"./data/hsapiens/andreev_genes.csv")
```

```{r}
df_ensembl_hits <- read_csv("./data/hsapiens/runs/hits.csv")
library(ggplot2)

ggplot(df_ensembl_hits, aes(x=df_ensembl_hits)) + geom_histogram(bins = 30, color="black", fill="blue") +
  labs(title="Number of  gene repetitions in Ensembl CDS data, for genes in D.E. Andreev filtered set",
         subtitle="Exceeds FDR 10% threshold for ribo-seq",
         x = "Number of repetitions",
         y = "Number of genes")
```

```{r}
library(dplyr)

# Helper function for taking the right substring in R (why isn't this built in...)
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

# Reads in the dataset generated by read_fasta.py
andreev_fasta_df <- read_csv("./data/hsapiens/Homo_sapiens.cds.andreev.fa", col_names=F)

set.seed(1)
andreev_df <- andreev_fasta_df %>%
  # Ensures that the sequence starts with ATG and ends with a valid stop codon
  filter(substr(X2, 1, 3) == "ATG") %>%
  filter(substrRight(X2, 3) %in% c("TAG", "TAA", "TGA")) %>%
  # Randomly samples one member of each represented gene
  group_by(X3) %>%
  sample_n(size = 1)
colnames(andreev_df) <- c("FASTAHeader", "Sequence", "geneSymbol")

# Helper function for counting number of occurrences of character in string
strcount <- function(x, pattern, split=""){
unlist(lapply(
    strsplit(x, split),
       function(z) na.omit(length(grep(pattern, z)))
   ))
}
```

```{r}
andreev_df <- andreev_df %>%
  mutate(num_char = nchar(Sequence),
         countA = strcount(Sequence, "A"),
         countC = strcount(Sequence, "C"),
         countT = strcount(Sequence, "T"),
         countG = strcount(Sequence, "G"))

freq_A = sum(andreev_df$countA) / sum(andreev_df$num_char)
freq_T = sum(andreev_df$countT) / sum(andreev_df$num_char)
freq_C = sum(andreev_df$countC) / sum(andreev_df$num_char)
freq_G = sum(andreev_df$countG) / sum(andreev_df$num_char)

ntides = c("A","T","C","G")
frequencies = c(freq_A, freq_C, freq_T, freq_G)
freq_df <- data.frame(ntides, frequencies)

library(ggplot2)

ggplot(freq_df, aes(x=ntides, y=frequencies, fill=ntides)) + geom_bar(stat = "identity")
```

```{r}

```